cloudinary-image-management-and-manipulation-in-the-cloud-cdn/php/sync/class-storage.php

```diff
@@ -0,0 +1,242 @@
+	 *
+	 * @since   0.1
+	 *
+	 * @var     \Cloudinary\Connect Instance of the plugin.
+	 */
+	protected $connect;
+
+	/**
+	 * Holds an array of the storage settings.
+	 *
+	 * @var array
+	 */
+	protected $settings;
+
+	/**
+	 * Filter constructor.
+	 *
+	 * @param \Cloudinary\Plugin $plugin The plugin.
+	 */
+	public function __construct( \Cloudinary\Plugin $plugin ) {
+		$this->plugin   = $plugin;
+		$this->sync     = $this->plugin->components['sync'];
+		$this->media    = $this->plugin->components['media'];
+		$this->download = $this->sync->managers['download'];
+
+		add_action( 'cloudinary_register_sync_types', array( $this, 'setup' ), 20 );
+	}
+
+	/**
+	 * Generate a signature for this sync type.s
+	 *
+	 * @return string
+	 */
+	public function generate() {
+		return $this->settings['offload'] . $this->settings['low_res'];
+	}
+
+	/**
+	 * Process the storage sync for an attachment.
+	 *
+	 * @param int $attachment_id The attachment ID.
+	 */
+	public function sync( $attachment_id ) {
+
+		switch ( $this->settings['offload'] ) {
+			case 'cld':
+				$this->remove_local_assets( $attachment_id );
+				break;
+			case 'dual_low':
+				$url = $this->media->cloudinary_url( $attachment_id, 'full', array( array( 'effect' => 'blur:100', 'quality' => $this->settings['low_res'] . ':440' ) ), null, false, true );
+				break;
+			case 'dual_full':
+				$state = $this->media->get_post_meta( $attachment_id, Sync::META_KEYS['storage'], true );
+				if ( ! empty( $state ) && 'dual_full' !== $state ) {
+					// Only do this is it's changing a state.
+					$url = $this->media->cloudinary_url( $attachment_id, '', array(), null, false, false );
+				}
+				break;
+		}
+
+		// If we have a URL, it means we have a new source to pull from.
+		if ( ! empty( $url ) ) {
+			$this->remove_local_assets( $attachment_id );
+			$this->download->download_asset( $attachment_id, $url );
+		}
+
+		$this->sync->set_signature_item( $attachment_id, 'storage' );
+		$this->sync->set_signature_item( $attachment_id, 'breakpoints' );
+		$this->media->update_post_meta( $attachment_id, Sync::META_KEYS['storage'], $this->settings['offload'] ); // Save the state.
+	}
+
+	/**
+	 * @param $attachment_id
+	 *
+	 * @return bool
+	 */
+	protected function remove_local_assets( $attachment_id ) {
+		// Delete local versions of images.
+		$meta = wp_get_attachment_metadata( $attachment_id );
+
+		return wp_delete_attachment_files( $attachment_id, $meta, array(), get_attached_file( $attachment_id ) );
+	}
+
+	/**
+	 * Get the current status of the sync.
+	 *
+	 * @return string
+	 */
+	public function status() {
+		$note = __( 'Syncing', 'cloudinary' );
+		switch ( $this->settings['offload'] ) {
+			case 'cld':
+				$note = __( 'Removing local copies.', 'cloudinary' );
+				break;
+			case 'dual_low':
+				$note = __( 'Reducing local resolutions', 'cloudinary' );
+				break;
+			case 'dual_full':
+				$note = __( 'Rebuilding local copies.', 'cloudinary' );
+				break;
+		}
```

Comment:
What if we want to add a new option in the future? That'd mean that we have to find all these switch statements to add logic to. As a minimal solution, we could have an array as a class property with the offload options, something like this:

```
protected $offload_options = [
    'cld'       => [
      'handler' => function ( $attachment_id ) {
        $this->remove_local_assets( $attachment_id );
      },
      'note'    => __( 'Removing local copies.', 'cloudinary' )
    ],
    'dual_low   => [
      'handler' => function ( $attachment_id ) {
        return $this->media->cloudinary_url( $attachment_id, 'full', array( array( 'effect' => 'blur:100', 'quality' => $this->settings['low_res'] . ':440' ) ), null, false, true );
      },
      'note'    => __( 'Reducing local resolutions', 'cloudinary' );
    ],
    'dual_full' => [
      'handler' => function ( $attachment_id ) {
        $state = $this->media->get_post_meta( $attachment_id, Sync::META_KEYS['storage'], true );

	if ( ! empty( $state ) && 'dual_full' !== $state ) {
		// Only do this is it's changing a state.
		return $this->media->cloudinary_url( $attachment_id, '', array(), null, false, false );
	}

        return null;
      },
      'note'    => __( 'Rebuilding local copies.', 'cloudinary' )
    ]
];
```

Then you can simple use it as:

```
$url = $this->offload_options[ $this->settings['offload'] ]['handler']( $attachment_id ); // use call_user_func here instead
```

or 

```
$note = $this->offload_options[ $this->settings['offload'] ]['note'];
```

Like said, this is the minimal extensible method that I could think of. Ideally, each option would be its own class, but that's a whole topic on its own 😄.

What do you think?

Reply:
@dugajean - that's a great point... I initially wanted this to be an offloading hook were anyone can hook into and offload how they want.
But due to timeframe I had to hard code it.

I think for now, we should focus on the ticket, and totally link this up with extending it later. Specifically CLOUD-366